File: app/src/main/java/com/readrops/app/ReadropsApp.java
package com.readrops.app;

import android.annotation.SuppressLint;
import android.app.Application;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.os.Build;

import androidx.appcompat.app.AppCompatDelegate;
import androidx.preference.PreferenceManager;

import com.readrops.app.utils.SharedPreferencesManager;

import io.reactivex.plugins.RxJavaPlugins;

@SuppressLint("Registered")
public class ReadropsApp extends Application {

    public static final String FEEDS_COLORS_CHANNEL_ID = "feedsColorsChannel";
    public static final String OPML_EXPORT_CHANNEL_ID = "opmlExportChannel";
    public static final String SYNC_CHANNEL_ID = "syncChannel";

    @Override
    public void onCreate() {
        super.onCreate();

        RxJavaPlugins.setErrorHandler(e -> {
        });

        createNotificationChannels();

        PreferenceManager.setDefaultValues(this, R.xml.preferences, false);

        if (Boolean.valueOf(SharedPreferencesManager.readString(this, SharedPreferencesManager.SharedPrefKey.DARK_THEME)))
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
        else
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
    }

    private void createNotificationChannels() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel feedsColorsChannel = new NotificationChannel(FEEDS_COLORS_CHANNEL_ID,
                    getString(R.string.feeds_colors), NotificationManager.IMPORTANCE_DEFAULT);
            feedsColorsChannel.setDescription(getString(R.string.get_feeds_colors));

            NotificationChannel opmlExportChannel = new NotificationChannel(OPML_EXPORT_CHANNEL_ID,
                    getString(R.string.opml_export), NotificationManager.IMPORTANCE_DEFAULT);
            opmlExportChannel.setDescription(getString(R.string.opml_export_description));

            NotificationChannel syncChannel = new NotificationChannel(SYNC_CHANNEL_ID,
                    getString(R.string.auto_synchro), NotificationManager.IMPORTANCE_DEFAULT);
            syncChannel.setDescription(getString(R.string.account_synchro));

            NotificationManager manager = getSystemService(NotificationManager.class);

            manager.createNotificationChannel(feedsColorsChannel);
            manager.createNotificationChannel(opmlExportChannel);
            manager.createNotificationChannel(syncChannel);
        }
    }
}

File: app/src/main/java/com/readrops/app/ReadropsApp.kt
package com.readrops.app

import android.app.Application
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationManagerCompat
import androidx.preference.PreferenceManager
import coil3.ImageLoader
import coil3.PlatformContext
import coil3.SingletonImageLoader
import coil3.disk.DiskCache
import coil3.disk.directory
import coil3.network.okhttp.OkHttpNetworkFetcherFactory
import coil3.request.crossfade
import com.readrops.api.apiModule
import com.readrops.app.util.CrashActivity
import com.readrops.app.util.FeverFaviconFetcher
import com.readrops.app.util.Migrations
import com.readrops.db.dbModule
import kotlinx.coroutines.runBlocking
import okhttp3.OkHttpClient
import org.koin.android.ext.android.get
import org.koin.android.ext.koin.androidContext
import org.koin.android.ext.koin.androidLogger
import org.koin.core.component.KoinComponent
import org.koin.core.context.startKoin
import org.koin.core.logger.Level
import kotlin.system.exitProcess

open class ReadropsApp : Application(), KoinComponent, SingletonImageLoader.Factory {

    override fun onCreate() {
        super.onCreate()

        if (!BuildConfig.DEBUG) {
            Thread.setDefaultUncaughtExceptionHandler { _, throwable ->
                val intent = Intent(this, CrashActivity::class.java).apply {
                    putExtra(CrashActivity.THROWABLE_KEY, throwable)
                    addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
                    addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
                }

                startActivity(intent)
                exitProcess(0)
            }
        }

        startKoin {
            androidLogger(Level.ERROR)
            androidContext(this@ReadropsApp)

            modules(apiModule, dbModule, appModule)
        }

        createNotificationChannels()

        runBlocking {
            Migrations.upgrade(
                appPreferences = get(),
                encryptedPreferences = get(),
                oldPreferences = PreferenceManager.getDefaultSharedPreferences(this@ReadropsApp),
                database = get(),
            )
        }
    }

    override fun newImageLoader(context: PlatformContext): ImageLoader {
        return ImageLoader.Builder(this)
            .components {
                add(OkHttpNetworkFetcherFactory(callFactory = {
                    val client = get<OkHttpClient>()
                    // custom shared Okhttp instance to avoid mixing
                    // authentication headers with basic image calls
                    client.newBuilder()
                        .build()
                }))

                add(FeverFaviconFetcher.Factory(get()))
            }
            .diskCache {
                DiskCache.Builder()
                    .directory(this.cacheDir.resolve("image_cache"))
                    .maximumMaxSizeBytes(1024 * 1024 * 100)
                    .build()
            }
            .crossfade(true)
            .build()
    }

    private fun createNotificationChannels() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val syncChannel = NotificationChannel(
                SYNC_CHANNEL_ID,
                getString(R.string.auto_synchro),
                NotificationManager.IMPORTANCE_LOW
            )
            syncChannel.description = getString(R.string.account_synchro)

            NotificationManagerCompat.from(this)
                .createNotificationChannel(syncChannel)
        }
    }

    companion object {
        const val SYNC_CHANNEL_ID = "syncChannel"
    }
}